tests/math.egg

```diff
@@ -0,0 +1,178 @@
+(datatype Math
+    (Diff Math Math)
+    (Integral Math Math)
+    
+    (Add Math Math)
+    (Sub Math Math)
+    (Mul Math Math)
+    (Div Math Math)
+    (Pow Math Math)
+    (Ln Math)
+    (Sqrt Math)
+    
+    (Sin Math)
+    (Cos Math)
+    
+    (Const Rational)
+    (Var String))
+
+(relation MathU (Math))
+(rule ((= e (Diff x y))) ((MathU e)))
+(rule ((= e (Integral x y))) ((MathU e)))
+(rule ((= e (Add x y))) ((MathU e)))
+(rule ((= e (Sub x y))) ((MathU e)))
+(rule ((= e (Mul x y))) ((MathU e)))
+(rule ((= e (Div x y))) ((MathU e)))
+(rule ((= e (Pow x y))) ((MathU e)))
+(rule ((= e (Ln x))) ((MathU e)))
+(rule ((= e (Sqrt x))) ((MathU e)))
+(rule ((= e (Sin x))) ((MathU e)))
+(rule ((= e (Cos x))) ((MathU e)))
+(rule ((= e (Const x))) ((MathU e)))
+(rule ((= e (Var x))) ((MathU e)))
+
+(relation evals-to (Math Rational))
+
+(rule ((= e (Const c))) ((evals-to e c)))
+(rule ((= e (Add a b)) (evals-to a va) (evals-to b vb))
+      ((evals-to e (+ va vb))))
+(rule ((= e (Sub a b)) (evals-to a va) (evals-to b vb))
+      ((evals-to e (- va vb))))
+(rule ((= e (Mul a b)) (evals-to a va) (evals-to b vb))
+      ((evals-to e (* va vb))))
+(rule ((= e (Div a b)) (evals-to a va) (evals-to b vb) (!= vb (rational 0 1)))
+      ((evals-to e (/ va vb))))
+(rule ((evals-to x vx)) ((union x (Const vx))))
+
+(relation is-const (Math))
+(rule ((evals-to a va)) ((is-const a)))
+
+(relation is-sym (Math))
+(rule ((= e (Var s))) ((is-sym e)))
+
+(relation is-not-zero (Math))
+(rule ((evals-to x vx)
+       (!= vx (rational 0 1)))
+      ((is-not-zero x)))
+
+(relation is-const-or-distinct-var-demand (Math Math))
+(relation is-const-or-distinct-var (Math Math))
+(rule ((is-const-or-distinct-var-demand v w)
+       (is-const v))
+      ((is-const-or-distinct-var v w)))
+(rule ((is-const-or-distinct-var-demand v w)
+       (= v (Var vv))
+       (= w (Var vw))
+       (!= vv vw))
+      (is-const-or-distinct-var v w))
+
+(rewrite (Add a b) (Add b a))
+(rewrite (Mul a b) (Mul b a))
+(rewrite (Add a (Add b c)) (Add (Add a b) c))
+(rewrite (Mul a (Mul b c)) (Mul (Mul a b) c))
+;; These two rules are not present in the original egg tests,
+;; but are necessary for the last check
+(rewrite (Add (Add a b) c) (Add a (Add b c)))
+(rewrite (Mul (Mul a b) c) (Mul a (Mul b c)))
+
+(rewrite (Sub a b) (Add a (Mul (Const (rational -1 1)) b)))
+(rewrite (Div a b) (Mul a (Pow b (Const (rational -1 1)))) :when ((is-not-zero b)))
+
+(rewrite (Add a (Const (rational 0 1))) a)
+(rewrite (Mul a (Const (rational 0 1))) (Const (rational 0 1)))
+(rewrite (Mul a (Const (rational 1 1))) a)
+
+;; These two rules can make the first goal time out
+;; (rule ((MathU a)) ((union a (Add a (Const (rational 0 1))))))
+;; (rule ((MathU a)) ((union a (Mul a (Const (rational 1 1))))))
```

Comment:
We discussed a fix for these in slack. These just need some additional guards to work.